service/history/workflow/cache/cache.go

```diff
@@ -418,6 +436,29 @@ func (c *CacheImpl) validateWorkflowID(
 	return nil
 }
 
+func (c *cacheItem) OnPut() {
+	wfKey := c.ctx.GetWorkflowKey()
+	err := c.finalizer.Register(wfKey.String(), func(ctx context.Context) error {
+		if err := c.ctx.Lock(ctx, workflow.LockPriorityHigh); err != nil {
+			return err
+		}
+		defer c.ctx.Unlock(workflow.LockPriorityHigh)
+		c.ctx.Clear()
+		return nil
+	})
+	if err != nil {
+		c.logger.Warn("finalizer failed to register callback", tag.Error(err))
+	}
+}
+
+func (c *cacheItem) OnEvict() {
+	wfKey := c.ctx.GetWorkflowKey()
+	err := c.finalizer.Deregister(wfKey.String())
+	if err != nil {
+		c.logger.Warn("finalizer failed to de-register callback", tag.Error(err))
+	}
+}
+
```

Comment:
There is not much we can do if any of these operations fail, other than logging it.